### PR TITLE
IoService

### PR DESCRIPTION
**Pull request by jerch**
PR of an IoService, that deals with different encoding aspects and the input flowcontrol.

Idea:
- all IO to the backend is byte based and can be directly read/written from/to the pty
- setEncoding gets applied to input and output data and should reflect the OS pty encoding
- encodings can be added as needed
- a decoder decodes bytes to our internal UTF32, an encoder encodes string data to the foreign encoding as bytes
- binary data from within the terminal gets directly sent, no output encoding will be applied

### CODE REVIEW

**Comment by jerch** [comment]
@Tyriar Done with the basic layout of the IoService (only a few TODOs left in the service itself). Would be good to get a first conceptual review before integrating this further in the terminal itself.

About encodings:
The encoding setting now applies to both input and output (imho separately setting this makes not much sense) and should reflect the pty app encoding. Missing encodings can be added with `addEncoding` (prolly usable by addons later on).

About IO in general:
- input: Input buffering has moved into the service, imho its best to deal with it in one place. The `write` method now deals with string raw bytes at once. This needs to be benchmarked as it might decrease performance.
`write` now applies the encoding to raw bytes, strings and utf-8 data are handled as before, only difference here - we have no explicit `writeUtf8` method anymore (as the name would be misleading).
- output: Data output can be grabbed as before with an `onData` event, but now as raw bytes with encoding correctly applied. Additionally if one really wants to grab only string data sent from the terminal, there is now a `onStringData`. For raw data there is a `onRawData` counterpart.

The next steps are to integrate this with `Terminal.ts` and do some benchmarks. I dont expect surprises here, still I am unsure about the `string | Uint8Array` merge for `write`.

**Comment by jerch** [comment]
Benchmark results with merged input types:
```
   Context "out-test/benchmark/test/benchmark/Terminal.benchmark.js"
      Context "Terminal: ls -lR /usr/lib"
         Context "write"
            Case "#1" : 10 runs - average throughput: 28.63 MB/s
         Context "writeUtf8"
            Case "#1" : 10 runs - average throughput: 33.37 MB/s
         Context "writeNew string"
            Case "#1" : 10 runs - average throughput: 29.84 MB/s
         Context "writeNew Utf8"
            Case "#1" : 10 runs - average throughput: 34.39 MB/s
```
The combined write in IoService is even slightly faster, no clue why. I think we are safe to go with this simpler input API.

**Comment by jerch** [comment]
This service is getting in shape, thus I removed the WIP flag.

@Tyriar Not sure what I am doing wrong here but it seems I cannot create a service that gets both normal args and other service args, it always fails here: https://github.com/xtermjs/xterm.js/blob/58e1b0db38fc4044055f2114b2dc7d1f4d48d27b/src/common/services/InstantiationService.ts#L71

Maybe this is just an interim issue, I currently need InputHandler in the service to do the parsing and the other services used by CoreService.triggerDataEvent to move the functionality, which does not work.

Furthermore I think we need to discuss my API proposal here, as it would change some basic interactions with xterm.js (for the better ofc :smile_cat:). Note that my proposals in #2326 are somewhat deprecated now as this here is the result of these thoughts and actual implementation.

**Comment by jerch** [comment]
@Tyriar Basically done with this PR, up for review/discussion.

File: typings/xterm.d.ts [hunk 1]
```diff
@@ -649,25 +686,23 @@ declare module 'xterm' {
...
+     * `data` can either be raw bytes from the pty or a string.
+     * Raw bytes will be decoded with the active encoding (default UTF-8),
+     * string data will always be decoded as UTF-16.
+     * `callback` is an optional callback that gets called once the data
+     * chunk was processed by the parser. Use this to implement
+     * a flow control mechanism so the terminal can keep up with incoming
+     * data. If the terminal falls to much behind data will be lost (>50MB).
      */
-    write(data: string): void;
+    write(data: string | Uint8Array, callback?: () => void): void;
```
Comment (Tyriar) [inline typings/xterm.d.ts]: I think this is good, I'm really concerned about the API getting too complicated/intimidating with all this encoding stuff so anything we can do to simplify great. I expect the perf hit won't be so bad.

File: typings/xterm.d.ts [hunk 2]
```diff
@@ -649,25 +686,23 @@ declare module 'xterm' {
...
-     * Writes text to the terminal.
-     * @param data The text to write to the terminal.
+     /**
+     * Write data to the terminal.
+     * `data` can either be raw bytes from the pty or a string.
+     * Raw bytes will be decoded with the active encoding (default UTF-8),
+     * string data will always be decoded as UTF-16.
+     * `callback` is an optional callback that gets called once the data
+     * chunk was processed by the parser. Use this to implement
+     * a flow control mechanism so the terminal can keep up with incoming
```
Comment (Tyriar) [inline typings/xterm.d.ts]: We probably shouldn't mention flow control yet until we know what the evenutal solution will be.

File: src/common/Platform.ts [hunk 3]
```diff
@@ -12,8 +12,9 @@ interface INavigator {
 // We're declaring a navigator global here as we expect it in all runtimes (node and browser), but
 // we want this module to live in common.
 declare const navigator: INavigator;
+declare const process: any;
 
-const isNode = (typeof navigator === 'undefined') ? true : false;
+export const isNode = (typeof navigator === 'undefined' && typeof process !== 'undefined') ? true : false;
```
Comment (Tyriar) [inline src/common/Platform.ts]: Just navigator wasn't sufficient?

File: src/common/services/Services.ts [hunk 4]
```diff
@@ -70,10 +71,75 @@ export interface ICoreService {
...
+  triggerStringDataEvent(data: string): void;
+
+  /**
+   * Send raw byte data as string type from within the terminal.
+   * `data` should be a string of codepoints in byte range (0-255),
+   * higher bits will be stripped ('binary' encoding).
+   * Output encoding as given in `setEncoding` is not applied.
+   * Grab the data with the `onRawData` event.
+   */
+  triggerRawDataEvent(data: string): void;
```
[the remark this answers is not in the document]
Comment (jerch) [reply src/common/services/Services.ts]: Nope not yet, the first use case for it would be the mouse report which just just got merged into master.

File: typings/xterm.d.ts [hunk 5]
```diff
@@ -411,13 +418,31 @@ declare module 'xterm' {
...
+     * The passed data is meant to be handled as bytestring,
+     * e.g. to be applied with 'binary' encoding.
      * @returns an `IDisposable` to stop listening.
      */
-    onData: IEvent<string>;
+    onRawData: IEvent<string>;
+
+    /**
+     * Event listener for data from the terminal.
+     * This event is a union of `onStringData` and `onRawData` as raw bytes with correctly
```
[the remark this answers is not in the document]
Comment (jerch) [reply typings/xterm.d.ts]: Yes - the idea behind this is to give integrators a chance to grab either unencoded or encoded versions of the data, while onData is kinda a funnel of encoded bytes:

```
  data to be sent from xterm.js to pty:

   real string data                  byte data
   (may contain unicode >255)        (as bytestring 0..255)

  +------------------+              +------------------+
  |                  |  unencoded   |                  |
  |   onStringData   |    events    |     onRawData    |
  |                  |              |                  |
  +------------------+              +------------------+
           |                                  |
           |                                  |
current    |            encoding step         | "binary"
(90% UTF8) |                                  |
           |                                  |
           |       +------------------+       |
           +------->                  <-------+
                   |      onData      |
                   |                  |
                   +------------------+
                            |
                            v
                       data to pty
                       (as bytes)
```

File: typings/xterm.d.ts [hunk 6]
```diff
@@ -506,6 +531,18 @@ declare module 'xterm' {
...
+     * Add an IO encoding to the terminal instance.
+     * IO encodings are applied to data input and output as follows:
+     * - raw byte input will be decoded with the active encoding
+     * - string input always decodes as UTF-16
+     * - outgoing data of `onData` will be encoded with the active encoding
+     * 
+     * See `Terminal.encodings` for installed encodings. Change
+     * `ITerminalOptions.encoding` to set the active encoding.
+     */
+    addEncoding(encoding: IEncoding): void;
```
[the remark this answers is not in the document]
Comment (jerch) [reply typings/xterm.d.ts]: I added this to the public API for a yet to come encoding addon, that could contain most common legacy encodings (see questions above) and give ppl a way to implement their "yet another ancient encoding" as well. 

The more fundamental question is whether xterm.js should care for legacy encodings at all, why not go with UTF8 everywhere. Well thats debatable. The question leads to "Who should ensure, that xterm.js encoding matches the pty encoding needs?" My initial stance here was - oh thats the integrators' responsibility. I think vscode bridges that already somehow, still I remember several issues regarding that (weird chars popping up and such). Most/Any of the other integrations dont care at all, I think neither hyper, terminus or fluentTerminal evaluate the pty encoding and translate between UTF8 and "the other encoding". Thats a problem, and all of them have unresolved issue reports regarding this.
My conclusion - lets do it as all other emulators do it - offer a setting for active encoding on API level. This lifts some burden from integrators (transcoding stuff forth and back), and those that didnt care so far kinda get it for free lol.

Note that those encoding problems mostly happen on windows these days, as most POSIX systems default to UTF-8 now. Windows is a mixed case here, seems some shells/apps spit out UTF16, others still some oldish cpXXX. This can be bridged with this PR pretty easily now: An integrator just needs to show some dropdown setting thingy with supported encodings. Whenever a user starts something with a foreign encoding he can just swap the active encoding. No more ssh/powershell/wsl encoding failures :smile:

File: typings/xterm.d.ts [hunk 7]
```diff
@@ -649,25 +686,23 @@ declare module 'xterm' {
...
-    writeln(data: string): void;
-
-    /**
-     * Writes UTF8 data to the terminal. This has a slight performance advantage
-     * over the string based write method due to lesser data conversions needed
-     * on the way from the pty to xterm.js.
-     * @param data The data to write to the terminal.
-     */
-    writeUtf8(data: Uint8Array): void;
+    writeln(data: string, callback?: () => void): void;
```
[the remark this answers is not in the document]
Comment (jerch) [reply typings/xterm.d.ts]: :+1:

File: typings/xterm.d.ts [hunk 8]
```diff
@@ -930,6 +971,51 @@ declare module 'xterm' {
     readonly width: number;
   }
 
+  /**
+   * xterm.js encoding interface.
+   * See common/input/Encodings.ts for implementation examples.
```
[the remark this answers is not in the document]
Comment (jerch) [reply typings/xterm.d.ts]: Right. Is a pointer to some addon code safe here (yet to come)? If not lets just remove it (the API docs should contain enough hints to get this done).

File: src/common/Platform.ts (hunk 3 above)
Comment (jerch) [reply src/common/Platform.ts]: Seems trying to distinct nodejs env from some browser engine context is very hard. I searched for it quite some time getting very different "this works for me" results. The pitfalls always go along these lines - never trust a simple property check on the global object, as some engines allow complete overwrites.

Newer browser engines make attempts to overwrite globals a NOOP now (thus requesting the property give still gives the real deal), but nodejs doesnt. Thus I added a sentinel that most nodejs envs hopefully would not change as they many modules rely on it - `process`.

**Comment by jerch** [comment]
> Ok there's a lot going on in this PR and I think we need to split it up:
> 
> Things I'm unsure about:
> 
>     * Anything that complicates the API (encodings, onData changes)
> 
>     * trigger(String|Raw)DataEvent
> 
>     * It's not clear to me why we need all these encoding and where they would be used

I have added an ASCII chart above to illustrate the usage of the trigger methods and the events further.

Practically we dont need other encodings in the codebase beside UTF-8, as >90% of all are UTF-8 now. The other <10% are either legacy systems ppl might ssh into, or windows related. Those fail atm. Since those encodings are not that common anymore, they are candidates for an addon.

> 
> Things we should get into v4:
> 
>     * The callback on write, I really want to start using this in both xterm.js tests and vscode tests. This could be a tiny PR merged very quickly.
> 
> 
> Things we should do in v4.1:
> 
>     * Introduce IOService without the external API changes first

Imho this separation does not work, as it does not address the problem with raw byte data vs. string data. To get support for this, we definitely gonna need an API change.

File: typings/xterm.d.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (Tyriar) [reply typings/xterm.d.ts]: I'd say remove it for now

File: typings/xterm.d.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (Tyriar) [reply typings/xterm.d.ts]: Actually it would be:

```ts
this.write(data);
this.write('\r\n', callback);
```

**Comment by jerch** [comment]
@Tyriar Now that v4 is out, we cannot apply the API changes anymore this PR would create. I still think that the encoding issues are a real problem, if you look through the integration projects like hyper, terminus and fluentTerminal - they all have open issues regarding this, most are unanswered though. Not sure why, not even sure if ppl know whats going on in this field.

Proposal to get this fixed (without breaking API change):
- rename `onStringData` in this PR back to `onData` as before
- rename `onData` in this PR to `onBytes` to reflect the byte nature of the data
- keep `onRawData` as bytestring event
- move all encodings into an addon, only provide UTF8 by default
- put encoding settings behind an API subsection
- fix attach addon accordingly

File: src/common/Platform.ts (hunk 3 above)
Comment (jerch) [reply src/common/Platform.ts]: Btw thats how emscripten tries to detect things:

```
// *** Environment setup code ***
var ENVIRONMENT_IS_NODE = typeof process === 'object' && typeof require === 'function';
var ENVIRONMENT_IS_WEB = typeof window === 'object';
var ENVIRONMENT_IS_WORKER = typeof importScripts === 'function';
var ENVIRONMENT_IS_SHELL = !ENVIRONMENT_IS_WEB && !ENVIRONMENT_IS_NODE && !ENVIRONMENT_IS_WORKER;
```

**Comment by Tyriar** [comment]
TL;DR of our chat about encoding portion: Instead of supporting all encoding in xterm.js, support utf-8 and provide docs on how to configure it properly ($LANG, luit, etc.)

**Comment by jerch** [comment]
Gonna close this issue as it mostly obsolete now. The remaining parts (changes to write + onBinaryData event) will covered by 2 following PRs.
